core/src/main/java/org/elasticsearch/common/lucene/search/Queries.java

```diff
@@ -179,8 +179,6 @@ public static int calculateMinShouldMatch(int optionalClauseCount, String spec)
             result = calc < 0 ? result + calc : calc;
         }
 
-        return (optionalClauseCount < result ?
-                optionalClauseCount : (result < 0 ? 0 : result));
-
+        return result < 0 ? 0 : result;
```

Comment:
> I can modify ExtendedCommonTermsQuery if you think that it's clearer this way.

Actually I was more thinking that our query parsers should not replace negative values with "0". They should either propagate to the query and let them deal with it or throw an exception instead of being lenient?